### PR TITLE
Enable additional tests for MPS CI runs

### PR DESCRIPTION
**Pull request by skotapati**
As part of the follow up for https://github.com/pytorch/pytorch/issues/133520, adapting existing unused tests for use in MPS CI runs. Focusing on nhwc & other memory formatting tests

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

**Comment by skotapati** [comment]
Note that test_nn.py is also being looked at by https://github.com/pytorch/pytorch/pull/134184

File: test/nn/test_convolution.py [hunk 1]
```diff
@@ -1855,6 +1857,7 @@ def test_conv2d_same_padding_backward(self, device, dtype):
         self.assertEqual(gy_expect, y.grad)
 
     @dtypes(torch.double, torch.cdouble)
+    @skipMPS  # Double, complex double not supported on MPS
```
Comment (malfet) [inline test/nn/test_convolution.py:1860]: IMO it's better this way, isn't it?
```suggestion
    @dtypesIfMPS(torch.float, torch.cfloat)  # Double, complex double not supported on MPS
```

File: test/nn/test_convolution.py [hunk 2]
```diff
@@ -2102,6 +2106,7 @@ def test_conv3d_valid_padding_backward(self, device, dtype):
         )
 
     @parametrize_test("N", range(2, 4), name_fn=lambda N: f"ConvTranspose{N}d")
+    @skipMPS  # ConvTranspose 3D not supported on MPS
```
Comment (malfet) [inline test/nn/test_convolution.py:2109]:
```suggestion
    @expectedFailureMPS  # ConvTranspose 3D not supported on MPS
```

File: test/nn/test_dropout.py [hunk 3]
```diff
@@ -207,6 +208,7 @@ def _test_dropoutNd_channel_zero(self, dropout, input):
             self.assertTrue(result[b, c].count_nonzero() in (0, channel_numel))
 
     @expectedFailureXLA  # seems like freeze_rng_state is not honoured by XLA
+    @expectedFailureMPS  # float64 not supported
```
Comment (malfet) [inline test/nn/test_dropout.py:211]: Why not use dtypesIfMPS?

File: test/nn/test_pooling.py [hunk 4]
```diff
@@ -818,6 +820,7 @@ def test_AvgPool2d_empty(self, device):
             inp = torch.randn(16, 0, 20, 32, device=device)
             avgpool(inp)
 
+    @skipIfMps  # max_pool3d_with_indices not supported on MPS
```
Comment (malfet) [inline test/nn/test_pooling.py:823]:
```suggestion
    @expectedFailureMPS  # max_pool3d_with_indices not supported on MPS
```

**Review by malfet — CHANGES_REQUESTED** [review]
Please do not use `skipIfMPS` unless there is a hard crash, if so, please create an issue that tracks fixing hard crash

Also, instead of expectedFailures due to unsupported types, use `dtypesIfMPS`

File: test/test_nn.py [hunk 5]
```diff
@@ -8953,7 +8953,7 @@ def check_rnn_grads(rnn1, rnn2):
     @dtypes(torch.double)
     def test_BatchNorm_empty(self, device, dtype):
         mod = torch.nn.BatchNorm2d(3).to(device)
-        inp = torch.randn(0, 3, 2, 2, device=device, dtype=dtype)
+        inp = torch.randn(0, 3, 2, 2, device=device)
```
Comment (malfet) [inline test/test_nn.py:8956]: Please don't change that one, but introduce `dtypeIfMPS`

File: test/test_nn.py [hunk 6]
```diff
@@ -8609,7 +8609,7 @@ def test_ReplicationPad1d_large(self, device):
             self.assertEqual(x.grad[:, :, 0], g[:, :, : pl + 1].sum(-1))
             self.assertEqual(x.grad[:, :, -1], g[:, :, -pr - 1:].sum(-1))
 
-    @expectedFailureMPS  # TODO(hvaara): Investigate as possible bug.
+    @expectedFailureMPS  # ReplicationPad has correctness issues on MPS - ISSUE#
```
Comment (malfet) [inline test/test_nn.py:8612]: Please file an issue and reference it here

**Comment by skotapati** [comment]
Whoops sorry, seems some rebasing issue occurred, please disregard

File: test/nn/test_convolution.py [hunk 7]
```diff
@@ -1677,6 +1679,7 @@ def test_conv_double_backward_stride(self):
                 )
 
     @dtypes(torch.float, torch.cfloat)
+    @dtypesIfMPS(torch.float)  # Complex not supported on MacOS13
```
Comment (malfet) [inline test/nn/test_convolution.py:1682]: How about something like that?
```suggestion
    @dtypesIfMPS(*([torch.float] if product_version < 14.0 else [torch.float, torch.cfloat]))  # Complex not supported on MacOS13
```

**Review by malfet — APPROVED** [review]
Looks good to me, but consider adding complex types conditional on product version (perhaps one can create global `mps_types` variable and do `@dtypesIfMPS(*mps_types)`

File: test/nn/test_convolution.py (hunk 7 above)
Comment (skotapati) [reply test/nn/test_convolution.py:1682]: Ah thanks, I wasn't sure if conditioning datatypes on OS version was ok or not. Adding that now

**Review by eqy — APPROVED** [review]

File: test/run_test.py [hunk 8]
```diff
@@ -1417,7 +1417,16 @@ def get_selected_tests(options) -> List[str]:
...
+            "test_modules",
+            "nn/test_convolution",
+            "nn/test_dropout",
+            "nn/test_pooling",
+            "test_view_ops",
+            "test_nn",
+        ]
     else:
         # Exclude all mps tests otherwise
         options.exclude.extend(["test_mps", "test_metal"])
```
Comment (huydhn) [inline test/run_test.py:1432]: I think you want to exclude the same tests above here.  Otherwise, they will be run in both trunk and mps

**Review by huydhn — APPROVED** [review]

**Pull request closed without merging**